### PR TITLE
Added trigger support to run caiman analysis pipeline

### PR DESCRIPTION
**Pull request by tdincer**
This PR allows element-calcium-imaging to run the default `caiman` analysis pipeline. The `caiman` module has to be installed from our own fork (vathes) for now.

### CODE REVIEW

File: element_calcium_imaging/imaging.py [hunk 1]
```diff
@@ -151,7 +151,8 @@ def make(self, key):
             else:
                 raise NotImplementedError('Unknown method: {}'.format(method))
         elif task_mode == 'trigger':
-            method = (ProcessingTask * ProcessingParamSet * ProcessingMethod * scan.Scan & key).fetch1('processing_method') 
+            method = (ProcessingTask * ProcessingParamSet * ProcessingMethod * scan.Scan & key).fetch1('processing_method')
+            print('method:', method)
```
Comment (kabilar) [inline element_calcium_imaging/imaging.py:155]:
```suggestion
```

Comment (tdincer) [reply element_calcium_imaging/imaging.py]: Thank you @kabilar! Nice catch!

**Review by kabilar — CHANGES_REQUESTED** [review]
Thanks @tdincer!  This is great.  A few minor comments:

- Lets import what you have under `element_calcium_imaging/readers/__init__.py` from [element_interface.scanimage_utils](https://github.com/datajoint/element-interface/blob/main/element_interface/scanimage_utils.py).  And `element_calcium_imaging/readers/caiman_loader.py` from [element-interace.caiman_loader](https://github.com/datajoint/element-interface/blob/main/element_interface/caiman_loader.py).
- For consistency, perhaps the `run_caiman` module should be within `element-interface`?
- Prior to merging, lets also submit the pytests within `workflow-calcium-imaging`.
- Since the updated `dj.Diagram` is within pull request #57, should we remove from this pull request?
- I think that this change warrants an update to the version and changelog.

**Comment by tdincer** [comment]
> * Lets import what you have under `element_calcium_imaging/readers/__init__.py` from [element_interface.scanimage_utils](https://github.com/datajoint/element-interface/blob/main/element_interface/scanimage_utils.py).  And `element_calcium_imaging/readers/caiman_loader.py` from [element-interace.caiman_loader](https://github.com/datajoint/element-interface/blob/main/element_interface/caiman_loader.py).
Yeah, definitely! I didn't know they existed in the element-interface.

> * For consistency, perhaps the `run_caiman` module should be within `element-interface`?
I don't strongly disagree, but this will make the dependencies pasta (e.g. caiman will be imported from `element-interface`)

> * Prior to merging, lets also submit the pytests within `workflow-calcium-imaging`.
This has to be done very quickly. The test can be identical to suite2p trigger test - we run both pipelines on the same images.

> * Since the updated `dj.Diagram` is within pull request #57, should we remove from this pull request?
I'm not sure what's the best way to avoid this conflict.

File: element_calcium_imaging/imaging.py [hunk 2]
```diff
@@ -165,13 +165,36 @@ def make(self, key):
...
-                
+
                 suite2p.run_s2p(ops=suite2p_params, db=suite2p_paths)  # Run suite2p
 
                 _, imaging_dataset = get_loader_result(key, ProcessingTask)
                 suite2p_dataset = imaging_dataset
                 key = {**key, 'processing_time': suite2p_dataset.creation_time}
 
+            elif method == 'caiman':
+                from element_interface.run_caiman import run_caiman
```
Comment (tdincer) [inline element_calcium_imaging/imaging.py:176]: `run_caiman` is imported from `element_interface` now.

File: element_calcium_imaging/imaging.py [hunk 3]
```diff
@@ -165,13 +165,36 @@ def make(self, key):
...
+                tiff_files = [f.as_posix() for f in tiff_files]
+
+                params = (ProcessingTask * ProcessingParamSet & key).fetch1('params')
+                sampling_rate = (ProcessingTask * scan.Scan * scan.ScanInfo & key).fetch1('fps')
+                output_dir = (ProcessingTask & key).fetch1('processing_output_dir')
+                output_dir = find_full_path(get_imaging_root_data_dir(), output_dir).as_posix()
+
+                ndepths = (ProcessingTask * scan.Scan * scan.ScanInfo & key).fetch1('ndepths')
+
+                is3D = True if ndepths > 1 else False
```
Comment (dimitri-yatsenko) [inline element_calcium_imaging/imaging.py:189]:
```suggestion
                is3D = ndepths > 1
```

Comment (tdincer) [reply element_calcium_imaging/imaging.py:189]: @dimitri-yatsenko  I tried `ndepths > 1`before,  it returned `np.bool_` type, and `caiman` wasn't happy with that.

Comment (dimitri-yatsenko) [reply element_calcium_imaging/imaging.py:189]: Then `bool(ndepths > 1)`

File: element_calcium_imaging/imaging.py [hunk 4]
```diff
@@ -165,13 +165,36 @@ def make(self, key):
...
+                if is3D:
+                    raise NotImplementedError('Caiman pipeline is not capable of analyzing 3D scans at the moment.')
```
Comment (dimitri-yatsenko) [inline element_calcium_imaging/imaging.py:191]: Is this error handing really necessary? Can you just catch the exception raised by `run_caiman`?

File: element_calcium_imaging/imaging.py [hunk 5]
```diff
@@ -165,13 +165,36 @@ def make(self, key):
...
                 _, imaging_dataset = get_loader_result(key, ProcessingTask)
                 suite2p_dataset = imaging_dataset
                 key = {**key, 'processing_time': suite2p_dataset.creation_time}
 
+            elif method == 'caiman':
+                from element_interface.run_caiman import run_caiman
+
+                tiff_files = (ProcessingTask * scan.Scan * scan.ScanInfo * scan.ScanInfo.ScanFile & key).fetch('file_path')
+                tiff_files = [find_full_path(get_imaging_root_data_dir(), tiff_file) for tiff_file in tiff_files]
+                tiff_files = [f.as_posix() for f in tiff_files]
```
Comment (dimitri-yatsenko) [inline element_calcium_imaging/imaging.py:180]:
```suggestion
                tiff_files = [find_full_path(get_imaging_root_data_dir(), tiff_file).as_posix() for tiff_file in tiff_files]
```

File: element_calcium_imaging/imaging.py [hunk 6]
```diff
@@ -165,13 +165,36 @@ def make(self, key):
...
+                ndepths = (ProcessingTask * scan.Scan * scan.ScanInfo & key).fetch1('ndepths')
+
+                is3D = True if ndepths > 1 else False
+                if is3D:
+                    raise NotImplementedError('Caiman pipeline is not capable of analyzing 3D scans at the moment.')
+                run_caiman(file_paths=tiff_files, parameters=params, sampling_rate=sampling_rate, output_dir=output_dir, is3D=is3D)
+
+                _, imaging_dataset = get_loader_result(key, ProcessingTask)
+                caiman_dataset = imaging_dataset
+                key = {**key, 'processing_time': caiman_dataset.creation_time}
```
Comment (dimitri-yatsenko) [inline element_calcium_imaging/imaging.py:196]:
```suggestion
                key.update(processing_time=caiman_dataset.creation_time)
```

File: element_calcium_imaging/imaging.py (hunk 4 above)
Comment (tdincer) [reply element_calcium_imaging/imaging.py:190]: These are two comments from the `caiman` pipeline:
> TODO - motion_correct_3d needs to be implemented in movies.py
> motion_correct_3d has not been implemented yet - instead initialize to just median image

Additionally, when I tried running the caiman pipeline on a 3D scan data, the code neither throw an exception nor was able to complete the run in an hour for a very small dataset (I had to cancel the run). Asa a result I came to the conclusion that we should add this error handling and take a deeper look into this later.

**Review by ttngu207 — APPROVED** [review]
Looks great @tdincer. A few points I would consider for the next iteration:
+ Can we trigger suite2p or caiman for ScanBox? if not then another `if/else` is needed based on the `AcquisitionSoftware` (similar to the ephys side)
+ We should think of a "resumable" mechanism for the triggering - e.g. to handles cases where something crashes in between and the next job does have to start over. Or worse yet, the analysis has finished but connection to the DB dropped, hence it cannot do the final insert step. This job would error out and the next time it tries again, the whole analysis is started over.

Those are for next development points. This PR is ready to be merged in its current form. 
@kabilar feel free to merge when you see fit

**Comment by tdincer** [comment]
> * Can we trigger suite2p or caiman for ScanBox? if not then another `if/else` is needed based on the `AcquisitionSoftware` (similar to the ephys side)

@ttngu207 I think we can trigger suite2p on scanbox data. As far as I remember suite2p converts `.sbx` files into `.bin` files and then runs the pipeline. I'll get back to this comment after I confirm it.

> * We should think of a "resumable" mechanism for the triggering - e.g. to handles cases where something crashes in between and the next job does have to start over. Or worse yet, the analysis has finished but connection to the DB dropped, hence it cannot do the final insert step. This job would error out and the next time it tries again, the whole analysis is started over.

Good that you noted this here!

**Comment by tdincer** [comment]
@ttngu207 and @kabilar 
I changed the ERD from png to svg (see [this](https://github.com/datajoint/element-calcium-imaging/pull/58/commits/7a1d51c6055e1624b41acbb2d55004e38ab1f9dc) and [this](https://github.com/datajoint/element-calcium-imaging/pull/58/commits/19e7c545f0ab467f53b5e085f5fc356a3d4a78ea)). I think this PR is ready to merge.

**Comment by tdincer** [comment]
> * Can we trigger suite2p or caiman for ScanBox? if not then another `if/else` is needed based on the `AcquisitionSoftware` (similar to the ephys side)

@ttngu207 I tried running the workflow with the caiman pipeline on a scanbox data (subject3) but `ScanInfo` couldn't handle the data as the data is `multiroi`. This is where we got the error in `ScanInfo`:

> is_multiROI = bool(sbx_matinfo.mesoscope.enabled)  # currently not handling "multiROI" ingestion
> 
> if is_multiROI:
>     raise NotImplementedError(
>         'Loading routine not implemented for Scanbox multiROI scan mode')

Looks like caiman can handle scanbox data (see https://github.com/flatironinstitute/CaImAn/issues/587)

**Pull request merged**